modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java

```diff
@@ -660,7 +660,7 @@ private void addElements() {
     }
 
     private void addStruct(final String name, final Class<?> clazz) {
-        if (!name.matches("^[_a-zA-Z][\\.,_a-zA-Z0-9]*$")) {
+        if (!name.matches("^[_a-zA-Z][\\.,_a-zA-Z0-9\\$]*$")) {
```

Comment:
We actually don't allow types in Painless to have the '$' as part of the type.  If I remember correctly the history here is that we have internal variables that have '$' that we don't want to have conflicts.  See later comment for a recommendation on how to resolve this.

If we want to change this we'd also have to modify the existing lexer/parser which I'm not sure is worth it with the workaround I mention in a later comment.